src/rules/unique-dependencies.ts

```diff
@@ -19,43 +19,86 @@ const dependencyPropertyNames = new Set([
 	"peerDependencies",
 ]);
 
-// This is a list of groups that required uniqueness among all of them globally.
-const groupsRequiringGlobalUniqueness = new Set([
-	"dependencies",
-	"devDependencies",
-]);
-
-// This keep tracks of dependencies seen from above set.
 export const rule = createRule({
 	create(context) {
-		const globalUniqueDependenciesSeen = new Set();
+		// This is a list of groups that required uniqueness among all of them globally.
+		// Inside each object, groups defined the corresponding list of groups to check for.
+		// "remove" will determine in which dependency group should the duplicated dependency be removed.
+		// eg: if same dependency appears in "dependencies" and "deveDependencies" we keep that from "dependencies".
+		const groupsRequiringGlobalUniqueness = [
+			{
+				groups: ["dependencies", "devDependencies"],
+				remove: "devDependencies",
+				seen: new Map<string, JsonAST.JSONStringLiteral>(),
+			},
+		];
+
+		// This keep tracks of a reference of each dependency node. We would be able to remove from the desired node.
+		const dependencyPropertyMap = new Map();
 		function check(
 			dependencyGroup: string,
 			elements: (JsonAST.JSONNode | null)[],
-			getNodeToRemove: (element: JsonAST.JSONNode) => JsonAST.JSONNode,
 		) {
 			const seen = new Set();
 
 			for (const element of elements
 				.filter(isNotNullish)
 				.filter(isJSONStringLiteral)
 				.reverse()) {
-				if (groupsRequiringGlobalUniqueness.has(dependencyGroup)) {
-					if (globalUniqueDependenciesSeen.has(element.value)) {
-						report(element, elements);
-					} else {
-						globalUniqueDependenciesSeen.add(element.value);
-					}
-				} else if (seen.has(element.value)) {
-					report(element, elements);
+				// First do a simple check to see if this dependency is being duplicated inside its current group.
+				if (seen.has(element.value)) {
+					const { getNodeToRemove } =
+						dependencyPropertyMap.get(dependencyGroup);
+					report(element, elements, getNodeToRemove);
+					continue;
 				} else {
 					seen.add(element.value);
 				}
+
+				// Next, check if we need to check for global unique dependencies.
+				if (
+					groupsRequiringGlobalUniqueness.some((groupsList) =>
+						groupsList.groups.includes(dependencyGroup),
+					)
+				) {
+					// We will always need to iterate through every item in this case, as we would want to report every duplication.
+					groupsRequiringGlobalUniqueness.forEach(
```

Comment:
Wouldn't we only want to do this on group lists that include the current dependencyGroup?  The if condition is saying "there's at least one group list in my array that includes the dependency group", but then here, you're looping through the entire array, rather than just the group list that includes the dependency group.  This also results in a traversal of the list twice, when it could just be one for loop.  Recommend refactoring this to be a single for loop, which has an if to check if that group list includes the dependency group.  Also, recommend a traditional for loop instead of using `.forEach()` since the latter is generally less performant.